### DISCUSSION TITLE
Questions about TGen output

### DISCUSSION THREAD

**Discussion by timwalsh300** [shadow/shadow#2761] [Q&A] [answered]
In the stream-success TGen messages, can someone expound on the meaning of usecs-to-proxy-response and usecs-to-command? It looks like they're usually the same time, along with the time for first-byte-send. Are they always the same? From one stream to the next, what causes this time to vary (as I'm seeing it in the range of maybe 100 to 900 ms, just taking a casual glance at the output from my TGen clients, which send a small request every few seconds)?

**Comment by stevenengler** [comment, marked as answer]
I don't know the answers, but you might find [this table](https://github.com/shadow/tgen/issues/30) helpful.

> - usecs-to-proxy-response — "Tor says OK connection opened"
> - usecs-to-command — "TGen client sends (through Tor) command to tell the TGen server what to do"
> - usecs-to-first-byte-send — "Client sent first byte of transfer"

Are you using tgen with tor? If so, `usecs-to-proxy-response` is probably when tgen receives a "request granted" SOCKS reply from tor, which would be after tor has created a new stream (has chosen or generated a circuit, and has used the circuit to connect to the destination). The amount of time this takes depends on your network and what relays tor chooses to build the circuit through, and if it already has a pool of pending circuits to choose from.

**Reply by timwalsh300** [reply]
That's very helpful, thank you. Yes, the clients are using Tor.

> depends on... what relays tor chooses to build the circuit through

That's the key thing that I'm wondering. Like if I want to judge the performance of a given circuit, should I penalize it when the client reports a higher usecs-to-first-byte-send because that depends on the circuit, or should I ignore or subtract this time from the other reported times because it's independent of the circuit? My intuition was the latter. I guess I need to learn more about SOCKS to understand why it might be the former.

**Reply by stevenengler** [reply]
Does "performance" here mean throughput? Are you trying to judge the circuit before the stream starts, or can you use the stream data/timings to judge the circuit? I don't think `usecs-to-first-byte-send` would be a good metric to judge a circuit, but I'm not really sure what you're trying to do.

**Reply by timwalsh300** [reply]
Yeah, part of what I'm exploring is possible definitions of "performance" for different purposes. The idea is to use stream data/timings to judge circuits, and then see if that yields novel insights about what characteristics might make a given circuit "better" or "worse."

If usecs-to-first-byte-send is partially determined by a circuit's responsiveness when a client wants to initiate a stream, then it might be a useful metric, maybe even by itself. If it's only determined by things like the load on the client's CPU or how many streams the client is trying to initiate, then maybe I should normalize all the other times (like usecs-to-last-byte-recv) for that stream by subtracting it. By analogy, let's say I want to measure the performance of a fast food restaurant by the time it takes to get my food after I walk in the door. I'm imagining that first-byte-send is when I start placing my order. If that includes the time that I had to stand in line before placing my order, that's meaningful. But if it mostly includes time that I spent distracted by my phone after walking inside while the staff waited for me to place my order, then I should subtract that from the total time (and really just measure from last-byte-send to last-byte-recv).

Anyway, it looks like this is something that I need to understand better about Tor vs. something about Shadow at this point.
